### PR TITLE
Add LP-MPPI low-pass perturbation sampling

### PR DESCRIPTION
**Pull request by mohamedsamirx**
## Basic Info

  | Info | Please fill out this column |
  | ------ | ----------- |
  | Ticket(s) this addresses | Follow-up / replacement for #5997 |
  | Primary OS tested on | Ubuntu 24.04, ROS 2 Rolling container |
  | Robotic platform tested on | TurtleBot3-style Nav2 loopback simulation, plus RViz replay of recorded headless traces |
  | Does this PR contain AI generated software? | Yes |
  | Was this PR description generated by AI software? | No |

  ---

  ## Description of contribution in a few bullet points

  - Adds optional LP-MPPI-style low-pass filtering of MPPI sampled perturbations,  This feature follows the LP-MPPI idea of filtering sampled control perturbations before rollout, based on the [LP-MPPI paper](https://arxiv.org/abs/2503.11717) The implementation keeps the change localized to MPPI noise generation and leaves the existing Nav2 controller workflow unchanged. The filter is disabled by default and can be enabled through parameters.
  - Adds three static MPPI parameters:
    - `use_low_pass_filter`
    - `filter_cutoff_frequency`
    - `filter_order`
  - The filter is applied inside the MPPI noise generator, before sampled controls are added to the nominal control sequence.
  - The implementation uses a digital Butterworth low-pass filter.
  - The feature is disabled by default, so existing MPPI behavior is unchanged unless explicitly enabled.
  - Adds warm-up-and-crop filtering to avoid suppressing the first samples in the MPPI horizon from zero IIR filter state.
  - Adds unit tests covering:
    - LP filtering reduces temporal perturbation variation.
    - Holonomic `vy` filtering path.
    - cutoff clamping near Nyquist.
    - invalid cutoff/order/model-dt handling.
    - filter coefficient design rejection.
    - near-term perturbation authority after warm-up-and-crop.

  ---

  ## Description of documentation updates

  - Updated `nav2_mppi_controller/README.md` with the new LP-MPPI parameters.
  - Updated default bringup params with disabled-by-default example values:
    - `use_low_pass_filter: false`
    - `filter_cutoff_frequency: 2.0`
    - `filter_order: 2`
  - docs.nav2.org should add the new parameters to MPPI controller documentation.
  - MPPI tuning guide should mention the observed cutoff/order tradeoff:
    - lower cutoff / higher order improves perturbation smoothness,
    - but can reduce responsiveness in tight terminal maneuvers.

  ---

  ## Description of how this change was tested

  ### Build and unit tests

  Test environment:

  - Ubuntu 24.04 host.
  - ROS 2 Rolling container.

  Commands used:

  ```bash
  source /opt/ros/rolling/setup.bash
  source /opt/overlay_ws/install/setup.bash

  MAKEFLAGS=-j1 CMAKE_BUILD_PARALLEL_LEVEL=1 \
  colcon build \
    --packages-select nav2_mppi_controller \
    --executor sequential \
    --parallel-workers 1 \
    --cmake-args -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_TESTING=ON
```
  Unit test results:

  noise_generator_test: 11/11 passed
  optimizer_unit_tests: 16/16 passed

  Additional repository checks:

  diff --check passed
  no conflict markers found

  ### Simulation / benchmark method

  I used Nav2 loopback simulation rather than live RViz for benchmark numbers.

  Reason:

  - Live RViz was too heavy on the test laptop and changed runtime behavior.
  - For metrics, Nav2 was run headless so controller timing was less affected by visualization.
  - Visual proof was generated by replaying saved traces in RViz after the headless run.

  Benchmark workflow:

  1. Launch Nav2 loopback simulation headlessly.
  2. Run a scenario/profile pair.
  3. Record:
      - final result JSON,
      - launch/evaluator logs,
      - command samples,
      - odometry samples,
      - optimal trajectory samples where available,
      - critic summaries where available.
  4. Compute metrics from recorded outputs.
  5. Replay saved trace in RViz only for visualization/screencast.

  This means benchmark numbers come from headless runs, while videos are visual replays of the exact saved traces.

  ### Scenarios tested

  Final selected matrix scenarios:

  - tb3_open_straight
  - tb3_turning_route
  - tb3_tight_s_turn
  - tb3_near_obstacle_crossing

  These were chosen to cover:

  - simple path following,
  - turning behavior,
  - tight terminal turning / progress-checking stress,
  - near-obstacle route following.

  Profiles in final selected matrix:

  - vanilla_sgf
  - lp_sgf_c4p0_o1
  - lp_sgf_c0p8_o1
  - related low-frequency profiles for comparison only, not included in this PR.

  | Profile | Meaning |
  |---|---|
  | `vanilla_sgf` | Baseline Nav2 MPPI with existing Savitzky-Golay smoothing only |
  | `lp_sgf_c4p0_o1` | LP-MPPI + Savitzky-Golay, cutoff `4.0 Hz`, order `1` |
  | `lp_sgf_c0p8_o1` | LP-MPPI + Savitzky-Golay, cutoff `0.8 Hz`, order `1` |

  Repeats:

  - 2 repeats per scenario/profile in the final selected matrix.

  ### Metrics computed

  The benchmark recorded and compared:

  - success count,
  - final distance to goal,
  - path length,
  - path efficiency,
  - command sample rate,
  - pose sample rate,
  - optimal trajectory sample rate,
  - mean/rms command deltas,
  - total variation,
  - acceleration metrics,
  - high-frequency energy ratio,
  - MSSD / MSGFD-style smoothness metrics,
  - inter-iteration perturbation/control variation,
  - critic summary statistics.

  The most relevant LP-MPPI metrics were:

  - optimal trajectory mean absolute dvx,
  - optimal trajectory mean absolute dwz,
  - inter-iteration dvx/dwz,
  - success rate,
  - final distance,
  - runtime / optimal trajectory publication rate.

  ### Final selected matrix result

  | Profile | Success | Mean final dist | Opt dvx | Opt dwz | Inter dvx | Inter dwz | Opt rate |
  |---|---:|---:|---:|---:|---:|---:|---:|
  | vanilla_sgf | 8/8 | 0.189 | 0.009025 | 0.008219 | 0.004821 | 0.010605 | 19.98 Hz |
  | lp_sgf_c4p0_o1 | 8/8 | 0.182 | 0.005628 | 0.009046 | 0.005238 | 0.013614 | 19.99 Hz |
  | lp_sgf_c0p8_o1 | 7/8 | 0.237 | 0.003227 | 0.007946 | 0.005052 | 0.014922 | 19.99 Hz |

  Interpretation:

  - lp_sgf_c4p0_o1 was the best balanced LP-MPPI candidate.
  - It matched vanilla reliability: 8/8.
  - It slightly improved mean final distance in this matrix: 0.182 m vs 0.189 m.
  - It reduced optimal trajectory vx variation: 0.005628 vs 0.009025.
  - It did not introduce observed runtime loss: 19.99 Hz vs 19.98 Hz.
  - It did not improve all smoothness axes: optimal dwz and inter-iteration metrics were similar or slightly worse than vanilla.
  - lp_sgf_c0p8_o1 was smoother on some optimal trajectory metrics, but less reliable: 7/8.

  ### Parameter tuning performed

  LP cutoff/order sweep:

  - cutoff: 0.8, 1.2, 2.0, 3.0, 4.0
  - order: 1, 2, 3, 4
  - scenario: tb3_tight_s_turn
  - repeats: 3 per profile

  Findings:

  - Lower cutoff and higher order produced smoother perturbations.
  - Very aggressive filtering commonly reduced responsiveness near the goal.
  - Conservative settings were more reliable.
  - Good candidates from tuning:
      - cutoff=3.0, order=4
      - cutoff=4.0, order=1
  - After the warm-up-and-crop correction and final selected matrix, cutoff=4.0, order=1 was the best balanced candidate.

  ### Failure / weakness analysis

  A key failure mode was observed with overly aggressive LP settings:

  - Scenario: tb3_tight_s_turn
  - Profile example: low cutoff / aggressive smoothing
  - Failure type: Nav2 progress failure near goal / status 6
  - Cause interpretation:
      - LP filtering reduces high-frequency exploration and command variation.
      - In tight terminal turns, too much filtering can reduce responsiveness near the goal.
      - The robot may get close to the goal but not satisfy progress checking fast enough.

  This is why the feature is disabled by default and why conservative tuning is recommended.

  ### Why warm-up-and-crop was added

  The initial causal IIR filtering approach started with zero filter state.

  Problem:

  - The first samples in the MPPI horizon were strongly suppressed.
  - Nav2 needs near-term control authority because the first part of the sampled horizon affects the immediate command.
  - Suppressing the first samples made tight turns weaker.

  Fix:

  - Generate extra noise samples before the real horizon.
  - Filter the longer sequence.
  - Crop the final time_steps samples back to the Nav2 horizon.

  This preserves the LP-MPPI idea while avoiding an implementation artifact from zero initial filter state.

  ### Visual validation

  Visual proof was done with trace replay:

  - First run headless Nav2 to generate exact trace.
  - Then replay the recorded trace in RViz.
  - This avoids changing benchmark behavior with RViz load.

  Visual cases prepared:

  - turning route, vanilla SGF
  - turning route, LP balanced setting
  - turning route, aggressive LP setting
  - tight S-turn, vanilla SGF
  - tight S-turn, LP balanced setting
  - tight S-turn, aggressive LP setting
  - near-obstacle crossing, vanilla SGF
  - near-obstacle crossing, LP balanced setting

  The visual evidence is intended to show both:

  - where LP-MPPI improves smoothness,
  - where over-filtering can reduce responsiveness.

  ### What this PR does not claim

  This PR does not claim LP-MPPI is a universal replacement for vanilla MPPI sampling.

  The observed claim is narrower:

  - Conservative LP-MPPI tuning can reduce sampled optimal trajectory jitter while preserving reliability and runtime in the tested loopback scenarios.
  - Aggressive LP filtering has a clear smoothness/responsiveness tradeoff.
  - The feature should remain opt-in.

  ### Related work not included in this PR

  I also evaluated low-frequency / colored-noise perturbation sampling separately.

  That work is intentionally not included here because:

  - it changes the sampling method differently from LP-MPPI,
  - it has different tuning parameters,
  - it showed different runtime and reliability tradeoffs,
  - it should be reviewed as a separate PR if pursued.

  Low-frequency findings will be discussed separately.

  ———

 I also prepared a small evidence folder with the selected benchmark table, replay instructions, visual recordings, and the matching JSON/log/trace artifacts for the LP-MPPI cases discussed above: https://drive.google.com/drive/folders/1Uc-durJ76RmLbV0ER7LsJWuZ752Mm3bj?usp=drive_link

  It includes only the referenced LP-MPPI / baseline visual cases and matrix summaries, not the full raw benchmark workspace. If maintainers want to reproduce or inspect the benchmark harness itself, I can also upload the benchmark scripts separately.

  ## Future work that may be required in bullet points

  - Add the new LP-MPPI parameters to docs.nav2.org.
  - Add MPPI tuning guide notes for cutoff/order tradeoffs.
  - Test on real robot hardware.
  - Test additional robot models beyond TB3-style loopback simulation.
  - Test more tight terminal maneuvers and near-obstacle paths.
  - Consider whether recommended LP settings should be documented as examples, not defaults.
  - Continue separate evaluation of low-frequency / colored-noise MPPI in a different PR.

### CODE REVIEW

**Comment by mohamedsamirx** [comment]
@SteveMacenski I also prepared the low-frequency / colored-noise MPPI branch related to #5973.

  Should I open that PR now as a stacked PR against this LP-MPPI branch, so it only shows the low-frequency diff, or would you prefer that I wait until this PR is merged/closed and then rebase the low-frequency branch onto `main` before opening it?

  I kept the low-frequency work separate because it has different parameters, test results, and tradeoffs from LP-MPPI.

File: nav2_bringup/params/nav2_params.yaml [hunk 1]
```diff
@@ -157,6 +157,9 @@ controller_server:
       motion_model: "diff_drive"
       diff_drive:
         plugin: "mppi::DiffDriveMotionModel"
+      use_low_pass_filter: false
```
Comment (nirwester) [inline nav2_bringup/params/nav2_params.yaml]: I'd group these under a common namespace for better encapsulation, ex:
```
noise_low_pass_filter:
  active: false
  cutoff_frequency: 2.0
  order: 2
```

File: nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp [hunk 2]
```diff
@@ -111,6 +130,16 @@ class NoiseGenerator
   std::condition_variable noise_cond_;
   std::mutex noise_lock_;
   bool active_{false}, ready_{false}, regenerate_noises_{false};
+
+  // Optional LP-MPPI sampling filter configuration.
+  bool use_low_pass_filter_{false};
```
Comment (nirwester) [inline nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp]: Could it make sense to implement the filter as a dedicated class, exposing methods for configuration and filtering?

File: nav2_mppi_controller/src/noise_generator.cpp [hunk 3]
```diff
@@ -78,6 +91,10 @@ void NoiseGenerator::reset(mppi::models::OptimizerSettings & settings, bool is_h
 {
   settings_ = settings;
   is_holonomic_ = is_holonomic;
+  ndistribution_vx_ = std::normal_distribution(0.0f, settings_.sampling_std.vx);
+  ndistribution_vy_ = std::normal_distribution(0.0f, settings_.sampling_std.vy);
+  ndistribution_wz_ = std::normal_distribution(0.0f, settings_.sampling_std.wz);
+  configureLowPassFilter();
```
Comment (nirwester) [inline nav2_mppi_controller/src/noise_generator.cpp]: If the noise generator is configured to run on its own thread (param regenerate_noises=true), I think there might be a chance of this clearing lpf_a_ and lpf_b_ while they are being used.
A possible solution could involve protecting this with noise_lock_ too.

File: nav2_mppi_controller/src/noise_generator.cpp [hunk 4]
```diff
@@ -34,6 +40,13 @@ void NoiseGenerator::initialize(
 
   auto getParam = param_handler->getParamGetter(name);
   getParam(regenerate_noises_, "regenerate_noises", false);
+  getParam(use_low_pass_filter_, "use_low_pass_filter", false, ParameterType::Static);
+  getParam(
```
Comment (nirwester) [inline nav2_mppi_controller/src/noise_generator.cpp:41]: I'd get (and hence declare) the filter parameters only if the filter is enabled to avoid expanding the configuration space when the filter is not used.

File: nav2_mppi_controller/src/noise_generator.cpp [hunk 5]
```diff
@@ -118,4 +164,176 @@ void NoiseGenerator::generateNoisedControls()
...
+    RCLCPP_WARN(
+      logger_,
+      "Low-pass filter disabled: filter_order must be >= 1, got %d",
+      filter_order_);
+    return;
+  }
+
+  if (filter_cutoff_frequency_ >= nyquist) {
+    const double clamped = std::max(
+      std::nextafter(nyquist, 0.0),
```
Comment (nirwester) [inline nav2_mppi_controller/src/noise_generator.cpp]: If nyquist is positive, the max operator returns something slightly bigger than nyquist; is that intended, or should it be clamped to something > 0 and < nyquist?

File: nav2_mppi_controller/src/noise_generator.cpp [hunk 6]
```diff
@@ -118,4 +164,176 @@ void NoiseGenerator::generateNoisedControls()
...
+      nyquist * 0.99);
+    RCLCPP_WARN(
+      logger_,
+      "filter_cutoff_frequency %.6f Hz is >= Nyquist %.6f Hz. Clamping to %.6f Hz.",
+      filter_cutoff_frequency_, nyquist, clamped);
+    filter_cutoff_frequency_ = clamped;
```
Comment (nirwester) [inline nav2_mppi_controller/src/noise_generator.cpp]: Note: if model_dt is reconfigured, the clamped filter_cutoff_frequency_ won't change. This is quite a corner case, but handling it wouldn't hurt :)

File: nav2_mppi_controller/test/noise_generator_test.cpp [hunk 7]
```diff
@@ -135,6 +204,266 @@ TEST(NoiseGeneratorTest, NoiseGeneratorMain)
...
+
+  const double filtered_vy_diff = meanAbsTemporalDifference(filtered_state.cvy);
+  const double raw_vy_diff = meanAbsTemporalDifference(raw_state.cvy);
+
+  EXPECT_LT(filtered_vy_diff, raw_vy_diff);
+}
+
+TEST(NoiseGeneratorTest, LowPassFilterWarmupKeepsNearTermPerturbationAuthority)
+{
+  auto node = std::make_shared<nav2::LifecycleNode>("lp_warmup_node");
```
Comment (nirwester) [inline nav2_mppi_controller/test/noise_generator_test.cpp:274]: To avoid repeating a lot of code, you could use a test fixture, but extracting the filter in a dedicated class would likely simplify the tests too.

**Review by nirwester — CHANGES_REQUESTED** [review]

File: nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp [hunk 8]
```diff
@@ -111,6 +113,14 @@ class NoiseGenerator
   std::condition_variable noise_cond_;
   std::mutex noise_lock_;
   bool active_{false}, ready_{false}, regenerate_noises_{false};
+
+  // Optional LP-MPPI sampling filter configuration.
+  bool low_pass_filter_active_{false};
```
Comment (nirwester) [inline nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp]: Since these are not dynamically updated, I suppose they could be persisted inside the NoiseLowPassFilter class instead of making them attributes of the NoiseGenerator. An additional dedicated resetDt method could be used for the reconfiguration in the reset function that only updates the dt.

File: nav2_mppi_controller/test/noise_generator_test.cpp [hunk 9]
```diff
@@ -135,6 +193,307 @@ TEST(NoiseGeneratorTest, NoiseGeneratorMain)
   generator.shutdown();
 }
 
+TEST(NoiseGeneratorTest, LowPassFilterSmoothsPerturbations)
```
Comment (nirwester) [inline nav2_mppi_controller/test/noise_generator_test.cpp:196]: Now that you have the additional class, it should now be possible to split NoiseGenerator tests from specific NoiseLowPassFilter tests

**Review by nirwester — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
Before reviewing this in detail, can you show me some before and after plots like this blog post? https://opennav.org/news/mppi-low-acceleration/ I would like to visually inspect the noise of the velocities rather than only base on benchmark metrics. Metrics are great and needed, but sometimes can hide other issues I'd like to make sure we sanity check. 

@nirwester please close any review comments that the author completed that you agree are completed so I can skip those when I get to a review. 

> Aggressive LP filtering has a clear smoothness/responsiveness tradeoff.

I recently added a SGF option for a 1st order filter vs a 2nd order. What I found was that was exceptionally smoother but the robot was not able to make more sophisticated maneuvers around obstacles due to smoothing out too much of the trajectory's signal / shape. That begs a few questions:

* Are you seeing that here too?
* Did you compare your work with the 1st order SGF and see some favorable differences or similarities? If similar, then I might perfer not to include additional features which add complexity without a clear differentiator. If there's some positive trade offs, please let me know :-) 
* Do you feel that the trade off you describe is worthwhile? I only see a major improved difference in the `Opt dvx` metric unless i'm misreading it


> I also prepared the low-frequency / colored-noise MPPI branch related to https://github.com/ros-navigation/navigation2/issues/5973.

This to me is actually the more interesting thing since we're not post-processing to improve smoothness and rather building smoothness into the formulation from the sampling distribution. 

I think its good to separate them and discuss them one at a time since they're decoupled contributions. Does that PR need this work in it?

**Comment by mohamedsamirx** [comment]
Absolutely, I can prepare before/after plots in the style of that blog post and update this PR once they’re ready.

I’ll generate plots for the same scenarios used in the benchmark runs, showing at least:
  - executed `cmd_vel` `vx` / `wz`
  - first differences / acceleration-like deltas
  - optimal trajectory velocity samples where available
  - vanilla SGF vs LP-MPPI tuned settings
  - SGF order 1 comparison, since that is the most relevant baseline for
  “smoother but maybe too smooth”

That should make it easier to visually inspect whether LP-MPPI is removing useful trajectory signal or only high-frequency noise.

On the smoothness/responsiveness tradeoff: yes, I am seeing that. In the tight S-turn case, aggressive LP filtering reduced within-horizon optimal trajectory variation, but hurt reliability/progress. Vanilla SGF completed that case, while the stronger LP settings failed with progress / empty-plan recovery behavior. So I do not think LP-MPPI should be presented as a general replacement for SGF or enabled by default.

I have not yet done a clean visual comparison against the new 1st-order SGF option, and I agree that is necessary before claiming this is worth the additional feature complexity. From the current metrics, the clearest LP-MPPI improvement is in the optimal trajectory / perturbation smoothness metrics, not consistently in executed `cmd_vel`. If the plots show behavior very similar to SGF order 1 without a meaningful differentiator, then I agree this may not justify inclusion as a separate option.

My current view is that the tradeoff is only worthwhile if the visual plots show a distinct benefit from shaping the sampled perturbations before rollout, rather than just producing another way to smooth the final/control trajectory. I’ll add that comparison and report back with the plots before asking for detailed review.
 
For the low-frequency / colored-noise work: it does not need this LP-MPPI PR. It is decoupled and can be opened separately from a clean branch based on `main`. I agree that one is more directly about building smoothness into the sampling distribution rather than post-processing, so I’ll keep it separate and discuss it independently. @SteveMacenski

**Comment by SteveMacenski** [comment]
> On the smoothness/responsiveness tradeoff: yes, I am seeing that. In the tight S-turn case, aggressive LP filtering reduced within-horizon optimal trajectory variation, but hurt reliability/progress. Vanilla SGF completed that case, while the stronger LP settings failed with progress / empty-plan recovery behavior. So I do not think LP-MPPI should be presented as a general replacement for SGF or enabled by default.

This is one of the key reasons I asked - wouldn't you say that its important that a robot can do such things? Are these extreme, unusual maneuvers or relatively straight forward things a robot would realistically need to do?

I ask because if its the latter, this may not be worth continuing to peruse. If it reduces functionality so much that it no longer can handle essential tasks, then I would say that this direction (like SGF order 1) is not an appropriate trade off. It needs to be smooth and retain agility in order to be a full solution -- as much as it sucks to say given the time you've spent on this otherwise good looking work :(

**Comment by mohamedsamirx** [comment]
I generated the before/after inspection plots you asked for, using the same headless loopback traces and evaluator JSONs. I included default SGF order 2, SGF order 1, and the two LP-MPPI candidates.

The plots support your concern. The tight S-turn / near-goal correction is not an extreme maneuver; it is a realistic maneuver a robot should be able to complete. In that case, both LP-MPPI candidates make the optimized trajectory smoother, but both fail the navigation action while both SGF baselines succeed.

So my current conclusion is that this LP-MPPI version is not a good general Nav2 tradeoff. It improves some smoothness metrics, especially optimized trajectory `dvx`, but it can smooth out enough near-term control authority to lose agility. Compared with SGF order 1, LP-MPPI does not currently show a strong enough positive differentiator to justify the added complexity.

I’ll attach the plots for transparency. Unless we find a tuning that keeps tight-maneuver reliability while improving velocity noise, I’m comfortable pausing or closing this LP-MPPI PR.

Thank you for appreciating the work, but it's okay, I learned a lot along the way.

Please take a look at the plots and decide what is better for the library. Feel free to close it. I will continue working on the low frequency and aim to open it today or tomorrow.
If there are other issues or improvements you'd like me to inspect after we finish these, feel free to ping me anytime.

tb3_turning_route_review_overview




tb3_tight_s_turn_review_overview



lp_mppi_metric_summary


lp_mppi_cmd_delta_distribution

@SteveMacenski

**Comment by SteveMacenski** [comment]
I appreciate the analysis and follow up and I agree with your similar feedback after looking over the charts. 

This was really good work and I'm glad to have learned along side you that this perhaps isn't a great addition and in the future we should be cautious about using post-processing to smooth out results rather than trying to create smoother results from sampling, the optimization problem, or critics themselves so we don't filter out important signals. 

As such, I would opt to close this even though its well implemented. The research itself doesn't stand up to practical scrutiny and I believe that the paper probably never really was an appropriate trade off, but got the metrics they were looking for for a publication :/

**Comment by mohamedsamirx** [comment]
Thank you. Agree, let's go to the main one now, haha. #6151
